creator-node/src/snapbackSM.js

```diff
@@ -47,63 +58,308 @@ class SnapbackSM {
-    // First step here is to implement discovery provider query
+    this.log('------------------Process SnapbackSM Operation------------------')
+    if (this.audiusLibs == null) {
+      logger.error(`Invalid libs instance`)
+      return
+    }
+    // Attempt to initialize, if this creator node has not yet been registered on chain
+    // then no operations will be performed
+    if (!this.initialized) {
+      await this.initializeNodeIdentityConfig()
+      return
+    }
+    // 1.) Retrieve base information for state machine operations
+    let spInfo = await this.getSPInfo()
+    if (spInfo.spID === 0) {
+      this.log(`Invalid spID, recovering ${spInfo}`)
+      await this.recoverSpID()
+      return
+    }
+    // Retrieve users list for this node
+    let usersList = await this.getNodePrimaryUsers()
+
+    // Generate list of wallets to query clock number
+    let nodeVectorClockQueryList = {}
+
+    // Issue queries to secondaries for each user
+    await Promise.all(
+      usersList.map(
+        async (user) => {
+          let userWallet = user.wallet
+          let secondary1 = user.secondary1
+          let secondary2 = user.secondary2
+          if (!nodeVectorClockQueryList[secondary1]) { nodeVectorClockQueryList[secondary1] = [] }
+          if (!nodeVectorClockQueryList[secondary2]) { nodeVectorClockQueryList[secondary2] = [] }
+          nodeVectorClockQueryList[secondary1].push(userWallet)
+          nodeVectorClockQueryList[secondary2].push(userWallet)
+        }
+      )
+    )
+
+    // Process nodeVectorClockQueryList and cache user clock values on each node
+    let secondaryNodesToProcess = Object.keys(nodeVectorClockQueryList)
+    let secondaryNodeUserClockStatus = {}
+    await Promise.all(
+      secondaryNodesToProcess.map(
+        async (node) => {
+          secondaryNodeUserClockStatus[node] = {}
+          // TODO: Batch this too?
+          let walletsToQuery = nodeVectorClockQueryList[node]
+          let requestParams = {
+            baseURL: node,
+            method: 'post',
+            url: '/users/batch_clock_status',
+            data: {
+              'walletPublicKeys': walletsToQuery
+            }
+          }
+          let resp = await axios(requestParams)
+          let userClockStatusList = resp.data.users
+          // Process returned clock values from this secondary node
+          userClockStatusList.map((entry) => {
+            secondaryNodeUserClockStatus[node][entry.walletPublicKey] = entry.clock
+          })
+        }
+      )
+    )
+
+    // Issue syncs if necessary
+    // For each user in the initially returned usersList,
+    //  compare local primary clock value to value from secondary retrieved in bulk above
+    await Promise.all(usersList.map(
+      async (user) => {
+        let userWallet = user.wallet
+        let secondary1 = user.secondary1
+        let secondary2 = user.secondary2
+        let primaryClockValue = await this.getUserPrimaryClockValue(userWallet)
```

Comment:
this will make a db query for each wallet, would be good to change the function to take array of wallets and query all at once before processing

Reply:
this is a great point, will definitely handle rn